core/src/main/java/com/linecorp/armeria/internal/common/TimeoutScheduler.java

```diff
@@ -97,93 +216,248 @@ private void extendTimeoutNanos(long adjustmentNanos) {
             });
         }
+    }
+
+    private boolean unsafeSetTimeoutNanosFromNow(long newTimeoutNanos) {
+        ensureInitialized();
+        if (state == State.TIMED_OUT || !timeoutTask.canSchedule()) {
+            return false;
+        }
+
+        // Cancel the previously scheduled timeout, if exists.
+        unsafeClearTimeout(true);
+
+        final long passedTimeNanos = System.nanoTime() - firstExecutionTimeNanos;
+        timeoutNanos = LongMath.saturatedAdd(newTimeoutNanos, passedTimeNanos);
+
+        if (newTimeoutNanos <= 0) {
+            return true;
+        }
 
-        this.timeoutNanos = LongMath.saturatedAdd(passedTimeNanos, timeoutNanos);
+        state = State.SCHEDULED;
+        timeoutFuture = eventLoop.schedule(this::invokeTimeoutTask, newTimeoutNanos, TimeUnit.NANOSECONDS);
+        return true;
     }
 
     public void timeoutNow() {
-        final TimeoutController timeoutController = this.timeoutController;
-        if (timeoutController != null) {
+        if (initialized) {
             if (eventLoop.inEventLoop()) {
-                timeoutController.timeoutNow();
+                unsafeTimeoutNow();
             } else {
-                eventLoop.execute(timeoutController::timeoutNow);
+                eventLoop.execute(this::unsafeTimeoutNow);
             }
         } else {
-            addPendingTimeoutTask(TimeoutController::timeoutNow);
+            addPendingTimeoutTask(this::unsafeTimeoutNow);
         }
     }
 
-    public boolean isTimedOut() {
-        if (timeoutController == null) {
-            return false;
+    private void unsafeTimeoutNow() {
+        checkState(timeoutTask != null,
+                   "init(eventLoop, timeoutTask) is not called yet.");
+
+        if (!timeoutTask.canSchedule()) {
+            return;
+        }
+
+        switch (state) {
+            case TIMED_OUT:
+                return;
+            case INIT:
+            case INACTIVE:
+                invokeTimeoutTask();
+                return;
+            case SCHEDULED:
+                if (unsafeClearTimeout(false)) {
+                    invokeTimeoutTask();
+                }
+                return;
+            default:
+                throw new Error(); // Should not reach here.
         }
-        return timeoutController.isTimedOut();
     }
 
-    public void setTimeoutController(TimeoutController timeoutController, EventLoop eventLoop) {
-        requireNonNull(timeoutController, "timeoutController");
-        requireNonNull(eventLoop, "eventLoop");
-        checkState(this.timeoutController == null, "timeoutController is set already.");
-        this.timeoutController = timeoutController;
-        this.eventLoop = eventLoop;
+    private void addPendingTimeoutTask(Runnable pendingTimeoutTask) {
+        if (!pendingTimeoutTaskUpdater.compareAndSet(this, null, pendingTimeoutTask)) {
+            for (;;) {
+                final Runnable pendingTask = this.pendingTimeoutTask;
```

Comment:
Question: Do we need to check if `initialized` is set in this loop and act differently?

Reply:
That's a good question. Let me check the `initialized` flag and execute `pendingTimeoutTask` with the `eventLoop` immediately.